src/hotspot/share/runtime/continuation.cpp

```diff
@@ -1130,106 +1134,96 @@ class Freeze : public FreezeBase {
-    int sp_before; // the chunk's sp before the freeze, adjusted to point beyond the stack-passed arguments in the topmost frame
+    int chunk_start_sp; // the chunk's sp before the freeze, adjusted to point beyond the stack-passed arguments in the topmost frame
     if (chunk_available) { // LIKELY
-      DEBUG_ONLY(allocated = false;)
       DEBUG_ONLY(orig_chunk_sp = chunk->sp_address();)
 
       assert(is_chunk_available0, "");
 
-      sp_before = chunk->sp();
-
-      if (sp_before < chunk->stack_size()) { // we are copying into a non-empty chunk
+      if (chunk->sp() < chunk->stack_size()) { // we are copying into a non-empty chunk
         DEBUG_ONLY(empty = false;)
-        assert(sp_before < (chunk->stack_size() - chunk->argsize()), "");
+        assert(chunk->sp() < (chunk->stack_size() - chunk->argsize()), "");
         assert(*(address*)(chunk->sp_address() - frame::sender_sp_ret_address_offset()) == chunk->pc(), "");
 
-        sp_before += stack_argsize; // we overlap; we'll overwrite the chunk's top frame's callee arguments
-        assert(sp_before <= chunk->stack_size(), "");
+        chunk_start_sp = chunk->sp() + _cont.argsize(); // we overlap; we'll overwrite the chunk's top frame's callee arguments
+        assert(chunk_start_sp <= chunk->stack_size(), "");
 
-        chunk->set_max_size(chunk->max_size() + size - stack_argsize);
+        chunk->set_max_size(chunk->max_size() + cont_size - _cont.argsize());
 
-        intptr_t* const bottom_sp = stack_bottom - stack_argsize;
+        intptr_t* const bottom_sp = cont_stack_bottom - _cont.argsize();
         assert(bottom_sp == _bottom_address, "");
         assert(*(address*)(bottom_sp-frame::sender_sp_ret_address_offset()) == StubRoutines::cont_returnBarrier(), "");
         patch_chunk_pd(bottom_sp, chunk->sp_address());
         // we don't patch the pc at this time, so as not to make the stack unwalkable
       } else { // the chunk is empty
-        DEBUG_ONLY(empty = true;)
-        assert(sp_before == chunk->stack_size(), "");
+        chunk_start_sp = chunk->sp();
+
+        assert(chunk_start_sp == chunk->stack_size(), "");
 
-        chunk->set_max_size(size);
-        chunk->set_argsize(stack_argsize);
+        chunk->set_max_size(cont_size);
+        chunk->set_argsize(_cont.argsize());
       }
     } else { // no chunk; allocate
-      DEBUG_ONLY(empty = true; allocated = true;)
-
       assert(_thread->thread_state() == _thread_in_vm, "");
       assert(!is_chunk_available(top_sp), "");
       assert(_thread->cont_fastpath(), "");
 
-      chunk = allocate_chunk(size + ContinuationHelper::frame_metadata);
-      if (UNLIKELY(chunk == nullptr)) return false; // OOME
-      if (UNLIKELY(!_thread->cont_fastpath()
-                  || _barriers)) { // probably humongous
+      chunk = allocate_chunk(cont_size + ContinuationHelper::frame_metadata);
+      if (UNLIKELY(chunk == nullptr || !_thread->cont_fastpath() || _barriers)) { // OOME/probably humongous
```

Comment:
This is just a strange set of conditions.  Why would you allocate_chunk if  _thread->cont_fastpath() is false? or _barriers, whatever that means.  Really don't expect a function like allocate_chunk to have side effects.
Maybe in the latter two conditions, you still want to zero out the chunk in the tail and don't want to remove this.

Reply:
The variable _barriers is set by allocate_chunk, so it must be check afterwards.
If allocation safepoints and there is JVM TI qeury it can 'remove' us as available for fastpath.
If allocation fails due to OOME.

The init_chunk() should always zero enought.

Reply:
Ok, thanks for confirming.